### DISCUSSION TITLE
Colour image not generating when visualizing the tessellation file

### DISCUSSION THREAD

**Discussion by Anmol-Bakshi** [neperfepx/neper#903] [Q&A] [answered]
When I am using the following tessellation file "***tess
 **format
   3.4
 **general
   2 standard
 **cell
  10
  *id
   1 2 3 4 5 6 7 8 9 10
  *crysym
   cubic
  *seed
   1 763.428324241280 355.339412596368 0.000000000000 518.690741609767
   2 4688.053011048509 660.316641268778 0.000000000000 966.638137927077
   3 985.143130470352 965.630616515137 0.000000000000 884.380743750188
   4 2473.737205384104 17947.171372557168 0.000000000000 16520.643587583483
   5 5280.240251523189 409.971710674110 0.000000000000 1016.100728778798
   6 8995.892747951108 1294.909613899809 0.000000000000 210.000000000000
   7 789.022732085509 268.707323963920 0.000000000000 827.470179839253
   8 9390.748130786542 1400.272765963564 0.000000000000 1246.708149414431
   9 10625.325485856622 1511.339962590778 0.000000000000 1158.692124524104
  10 5816.646212574373 330.897284219283 0.000000000000 333.704288416172
  *ori
   euler-bunge:active
 -75.000000000000   42.600000000000   80.300000000000
  27.600000000000   17.800000000000   52.400000000000
-188.400000000000   21.300000000000   35.600000000000
-189.900000000000   38.100000000000   75.700000000000
-200.500000000000   45.800000000000   43.200000000000
 -17.100000000000   33.000000000000   56.900000000000
-140.500000000000   20.700000000000   28.900000000000
 110.200000000000   39.000000000000   40.900000000000
-161.200000000000   21.700000000000   60.700000000000
  71.000000000000   14.300000000000   72.500000000000
 **vertex
 22
   1  0.000000000000 0.000000000000 0.000000000000     0
   2  2831.817293447034 -0.000000000000 -0.000000000000     0
   3  -0.000000000000 1028.790798070107 -0.000000000000     0
   4  2831.822253093062 0.008279023232 -0.000000000000     0
   5  2831.822770697501 0.001618187724 -0.000000000000     0
   6  4157.277097530133 -0.000000000000 -0.000000000000     0
   7  2831.822933224423 -0.000000000000 -0.000000000000     0
   8  2942.060836517172 1337.004200853173 -0.000000000000     0
   9  4824.421576934346 1578.120424958634 -0.000000000000     0
  10  -0.000000000000 1594.903904548612 -0.000000000000     0
  11  0.000000000000 1960.000000000000 0.000000000000     0
  12  7065.179028339460 1947.068461513226 -0.000000000000     0
  13  6716.842701023940 1880.967273471113 -0.000000000000     0
  14  7096.114576977877 1960.000000000000 -0.000000000000     0
  15  7065.189740340687 1947.072657059705 -0.000000000000     0
  16  6439.559413495680 -0.000000000000 -0.000000000000     0
  17  7065.277640556251 1946.743245186708 -0.000000000000     0
  18  10224.750014508829 -0.000000000000 -0.000000000000     0
  19  10048.421083401079 1960.000000000000 -0.000000000000     0
  20  7647.745500536952 -0.000000000000 -0.000000000000     0
  21  12238.600000000000 0.000000000000 0.000000000000     0
  22  12238.600000000000 1960.000000000000 0.000000000000     0
 **edge
 31
   1  3 1 0
   2  1 2 0
   3  3 4 0
   4  4 5 0
   5  5 2 0
   6  7 6 0
   7  7 5 0
   8  4 8 0
   9  8 9 0
  10  9 6 0
  11  10 3 0
  12  10 8 0
  13  11 10 0
  14  14 11 0
  15  12 13 0
  16  13 9 0
  17  14 15 0
  18  15 12 0
  19  6 16 0
  20  13 16 0
  21  17 12 0
  22  15 17 0
  23  2 7 0
  24  20 18 0
  25  19 14 0
  26  19 18 0
  27  20 17 0
  28  21 22 0
  29  18 21 0
  30  22 19 0
  31  16 20 0
 **face
 10
   1 5 2 1 3 4 5
     5 -2 -1 3 4 5
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   2 6 6 7 5 4 8 9
     6 -6 7 -4 8 9 10
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   3 4 4 3 10 8
     4 -3 -11 12 -8
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   4 8 12 13 9 8 10 11 14 15
     8 15 16 -9 -12 -13 -14 17 18
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   5 4 16 6 9 13
     4 -19 -10 -16 20
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   6 3 17 12 15
     3 21 -18 22
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   7 3 2 5 7
     3 -5 -7 -23
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   8 6 19 18 20 17 15 14
     6 26 -24 27 -22 -17 -25
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   9 4 22 21 18 19
     4 -28 -29 -26 -30
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
  10 5 17 20 16 13 12
     5 -27 -31 -20 -15 -21
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
 **domain
  *general
   cube
  *vertex
   4
   1  0.000000000000 0.000000000000 0.000000000000 x0y0
      1 1
   2  12238.600000000000 0.000000000000 0.000000000000 x1y0
      1 21
   3  12238.600000000000 1960.000000000000 0.000000000000 x1y1
      1 22
   4  0.000000000000 1960.000000000000 0.000000000000 x0y1
      1 11
  *edge
   4
   1 2 2 1
     y0
     7 2 6 19 23 24 29 31
   2 2 3 2
     x1
     1 28
   3 2 4 3
     y1
     3 14 25 30
   4 2 1 4
     x0
     3 1 11 13
***end
", i am getting a colourful image showing different grains when visulizing.

Whereas, when i am using following tessellation file:
***tess
 **format
   3.4
 **general
   2 standard
 **cell
  10
  *id
   1 2 3 4 5 6 7 8 9 10
  *crysym
   cubic
  *seed
   1 763.428324241280 355.339412596368 0.000000000000 518.690741609767
   2 4688.053011048509 660.316641268778 0.000000000000 966.638137927077
   3 985.143130470352 965.630616515137 0.000000000000 884.380743750188
   4 2473.737205384104 17947.171372557168 0.000000000000 16520.643587583483
   5 5280.240251523189 409.971710674110 0.000000000000 1016.100728778798
   6 8995.892747951108 1294.909613899809 0.000000000000 210.000000000000
   7 789.022732085509 268.707323963920 0.000000000000 827.470179839253
   8 9390.748130786542 1400.272765963564 0.000000000000 1246.708149414431
   9 10625.325485856622 1511.339962590778 0.000000000000 1158.692124524104
  10 5816.646212574373 330.897284219283 0.000000000000 333.704288416172
  *ori
   euler-bunge:active
 -75.000000000000   42.600000000000   80.300000000000
  27.600000000000   17.800000000000   52.400000000000
-188.400000000000   21.300000000000   35.600000000000
-189.900000000000   38.100000000000   75.700000000000
-200.500000000000   45.800000000000   43.200000000000
 -17.100000000000   33.000000000000   56.900000000000
-140.500000000000   20.700000000000   28.900000000000
 110.200000000000   39.000000000000   40.900000000000
-161.200000000000   21.700000000000   60.700000000000
  71.000000000000   14.300000000000   72.500000000000
 **vertex
 22
   1  0.000000000000 0.000000000000 0.000000000000     0
   2  3630.007293447034 -0.000000000000 -0.000000000000     0
   3  -0.000000000000 1028.790798070107 -0.000000000000     0
   4  0.000000000000 1960.000000000000 -0.000000000000     0
   5  575.822770697501 675.001618187724 -0.000000000000     0
   6  2300.000097530133 1075.000000000000 -0.000000000000     0
   7  2395.000033224423 1960.000000000000 -0.000000000000     0
   8  3560.060836517172 345.004200853173 -0.000000000000     0
   9  4115.421576934346 1960.000000000000 -0.000000000000     0
  10  3895.000000000000 0.000000048612 -0.000000000000     0
  11  5460.000000000000 1960.000000000000 0.000000000000     0
  12  5445.179028339460 1660.068461513226 -0.000000000000     0
  13  5451.842701023940 1473.967273471113 -0.000000000000     0
  14  5542.114576977877 1466.000000000000 -0.000000000000     0
  15  5947.189740340687 1960.000000000000 -0.000000000000     0
  16  5422.559413495680 -0.000000000000 -0.000000000000     0
  17  9229.277640556251 1960.000000000000 -0.000000000000     0
  18  9232.750014508829 889.000000000000 -0.000000000000     0
  19  9009.421083401079 0.000000000000 -0.000000000000     0
  20  9897.745500536952 -0.000000000000 -0.000000000000     0
  21  12238.600000000000 1960.000000000000 0.000000000000     0
  22  12238.600000000000 0.000000000000 0.000000000000     0
 **edge
 31
   1  3 4 0
   2  1 3 0
   3  1 2 0
   4  4 7 0
   5  3 5 0
   6  1 5 0
   7  5 6 0
   8  6 7 0
   9  6 8 0
  10  2 8 0
  11  7 9 0
  12  8 9 0
  13  10 13 0
  14  9 11 0
  15  11 12 0
  16  12 13 0
  17  13 14 0
  18  14 15 0
  19  11 15 0
  20  2 10 0
  21  10 16 0
  22  14 16 0
  23  15 17 0
  24  16 19 0
  25  17 18 0
  26  18 19 0
  27  19 20 0
  28  18 20 0
  29  17 21 0
  30  21 22 0
  31  20 22 0
 **face
 10
   1 3 1 5 3
     3 6 5 2
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   2 5 3 5 6 7 4
     5 5 7 8 4 1
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   3 5 5 1 2 8 6
     5 6 3 10 9 7
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   4 4 6 8 9 7
     4 9 12 11 8
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   5 7 8 2 10 13 12 11 9
     7 10 20 13 16 15 14 12
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   6 5 11 12 13 14 15
     5 15 16 17 18 19
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   7 4 10 16 14 13
     4 21 22 17 13
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   8 6 15 14 16 19 18 17
     6 18 22 24 26 25 23
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   9 5 17 18 20 22 21
     5 25 28 31 30 29
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
  10 3 18 19 20
     3 27 28 26
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
 **domain
  *general
   cube
  *vertex
   4
   1  0.000000000000 0.000000000000 0.000000000000 x0y0
      1 1
   2  12238.600000000000 0.000000000000 0.000000000000 x1y0
      1 22
   3  12238.600000000000 1960.000000000000 0.000000000000 x1y1
      1 21
   4  0.000000000000 1960.000000000000 0.000000000000 x0y1
      1 4
  *edge
   4
   1 2 2 1
     y0
     6 3 20 21 24 27 31
   2 2 3 2
     x1
     1 30
   3 2 4 3
     y1
     6 29 23 19 14 11 4
   4 2 1 4
     x0
     2 1 2
***end
And, now when I am visualizing it, black and white image just showing the grain boundaries are appearing. How to reslve it?
The visualization command i have used is "neper -V myfile.tess -datacellcol id -print Image_11" in both cases. 

Both images are attached.
![Image_11](https://github.com/user-attachments/assets/883950a4-a675-44a1-a306-975b966cc6f7)
![Image_12](https://github.com/user-attachments/assets/1cad3bd7-03b0-4e94-90ae-ca800b865e06)

**Comment by rquey** [comment]
I can only get the first image from you description.

**Reply by Anmol-Bakshi** [reply]
when i am using this tessellation file:
***tess
 **format
   3.4
 **general
   2 standard
 **cell
  10
  *id
   1 2 3 4 5 6 7 8 9 10
  *crysym
   cubic
  *seed
   1 195.428324241280 615.339412596368 0.000000000000 657.690741609767
   2 1310.053011048509 1355.316641268778 0.000000000000 1811.638137927077
   3 1965.143130470352 465.630616515137 0.000000000000 1740.380743750188
   4 3500.737205384104 1075.171372557168 0.000000000000 1720.643587583483
   5 4713.240251523189 980.971710674110 0.000000000000 1345.100728778798
   6 5588.892747951108 1704.909613899809 0.000000000000 487.000000000000
   7 5283.022732085509 825.707323963920 0.000000000000 1527.470179839253
   8 8009.748130786542 980.272765963564 0.000000000000 3385.708149414431
   9 11398.825485856622 1045.339962590778 0.000000000000 2335.692124524104
  10 10363.646212574373 235.897284219283 0.000000000000 889.704288416172
  *ori
   euler-bunge:active
 -75.000000000000   42.600000000000   80.300000000000
  27.600000000000   17.800000000000   52.400000000000
-188.400000000000   21.300000000000   35.600000000000
-189.900000000000   38.100000000000   75.700000000000
-200.500000000000   45.800000000000   43.200000000000
 -17.100000000000   33.000000000000   56.900000000000
-140.500000000000   20.700000000000   28.900000000000
 110.200000000000   39.000000000000   40.900000000000
-161.200000000000   21.700000000000   60.700000000000
  71.000000000000   14.300000000000   72.500000000000
 **vertex
 22
   1  0.000000000000 0.000000000000 0.000000000000     0
   2  3630.007293447034 -0.000000000000 -0.000000000000     0
   3  -0.000000000000 1028.790798070107 -0.000000000000     0
   4  0.000000000000 1960.000000000000 -0.000000000000     0
   5  575.822770697501 675.001618187724 -0.000000000000     0
   6  2300.000097530133 1075.000000000000 -0.000000000000     0
   7  2395.000033224423 1960.000000000000 -0.000000000000     0
   8  3560.060836517172 345.004200853173 -0.000000000000     0
   9  4115.421576934346 1960.000000000000 -0.000000000000     0
  10  3895.000000000000 0.000000048612 -0.000000000000     0
  11  5460.000000000000 1960.000000000000 0.000000000000     0
  12  5445.179028339460 1660.068461513226 -0.000000000000     0
  13  5451.842701023940 1473.967273471113 -0.000000000000     0
  14  5542.114576977877 1466.000000000000 -0.000000000000     0
  15  5947.189740340687 1960.000000000000 -0.000000000000     0
  16  5422.559413495680 -0.000000000000 -0.000000000000     0
  17  9229.277640556251 1960.000000000000 -0.000000000000     0
  18  9232.750014508829 889.000000000000 -0.000000000000     0
  19  9009.421083401079 0.000000000000 -0.000000000000     0
  20  9897.745500536952 -0.000000000000 -0.000000000000     0
  21  12238.600000000000 1960.000000000000 0.000000000000     0
  22  12238.600000000000 0.000000000000 0.000000000000     0
 **edge
 31
   1  3 4 0
   2  1 3 0
   3  1 2 0
   4  4 7 0
   5  3 5 0
   6  1 5 0
   7  5 6 0
   8  6 7 0
   9  6 8 0
  10  2 8 0
  11  7 9 0
  12  8 9 0
  13  10 13 0
  14  9 11 0
  15  11 12 0
  16  12 13 0
  17  13 14 0
  18  14 15 0
  19  11 15 0
  20  2 10 0
  21  10 16 0
  22  14 16 0
  23  15 17 0
  24  16 19 0
  25  17 18 0
  26  18 19 0
  27  19 20 0
  28  18 20 0
  29  17 21 0
  30  21 22 0
  31  20 22 0
 **face
 10
   1 3 1 5 3
     3 6 5 2
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   2 5 3 5 6 7 4
     5 5 7 8 4 1
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   3 5 5 1 2 8 6
     5 6 3 10 9 7
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   4 4 6 8 9 7
     4 9 12 11 8
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   5 7 8 2 10 13 12 11 9
     7 10 20 13 16 15 14 12
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   6 5 11 12 13 14 15
     5 15 16 17 18 19
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   7 4 10 16 14 13
     4 21 22 17 13
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   8 6 15 14 16 19 18 17
     6 18 22 24 26 25 23
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
   9 5 17 18 20 22 21
     5 25 28 31 30 29
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
  10 3 18 19 20
     3 27 28 26
    -0.000000000000 -0.000000000000 -0.000000000000 -1.000000000000
     0 -1 0.000000000000 0.000000000000 0.000000000000
 **domain
  *general
   cube
  *vertex
   4
   1  0.000000000000 0.000000000000 0.000000000000 x0y0
      1 1
   2  12238.600000000000 0.000000000000 0.000000000000 x1y0
      1 22
   3  12238.600000000000 1960.000000000000 0.000000000000 x1y1
      1 21
   4  0.000000000000 1960.000000000000 0.000000000000 x0y1
      1 4
  *edge
   4
   1 2 2 1
     y0
     6 3 20 21 24 27 31
   2 2 3 2
     x1
     1 30
   3 2 4 3
     y1
     6 29 23 19 14 11 4
   4 2 1 4
     x0
     2 1 2
***end

I am getting this image as attached. How to resolve it and get colour grains showing different orientations?
![Image_13](https://github.com/user-attachments/assets/ca6ed91c-6d54-4a88-a671-2bb16594e7a3)

**Reply by rquey** [reply, marked as answer]
This looks like a custom tess file (not generated by Neper).  You'll have to find what's wrong with it and make sure that it follows the specifications provided in [Tessellation File (.tess)](https://neper.info/doc/fileformat.html#tessellation-file-tess).

**Reply by Anmol-Bakshi** [reply]
Actually, I have first generated the tessellation file in neper using command "neper -T -n from_morpho -morpho "size:file(size.txt),centroid:file(coordinates.txt)" -dim 2 -ori "file(orientations.txt,des=euler-bunge)" -oridescriptor "euler-bunge:active" -domain "square(12238.6,1960)" -o simulation6"

And then I have just edited the seed values, edges, vertices and faces coordinates. And the domain vertex and edge description and tried to visualize it and I am getting this image.

So, I have maintained the specifications as I have just edited the description of vertex, edge and face in the one generated by neper.
Can you help in figuring out to start with where I am going wrong? I have also attached the image of meshed visulised in neper using "neper -V abq_tet_2d.msh -datacellcol id -print Image_16".

Any suggestions/insights where I am going wrong will be helpful as I have tried my best to match the syntax in tessellation file.

![Image_16](https://github.com/user-attachments/assets/1b52a34f-9452-4d3c-8819-52cba4a09d63)

**Reply by rquey** [reply]
I don't see any negative edge identifier inside `**face`, so this is one thing. Edges must be properly oriented.

**Reply by Anmol-Bakshi** [reply]
Thanks a lot for your suggestion, I could rectify it and got image showing different colours for different grains.
Also, I have made some syntax rules from my understanding which might be helpful to others:
**edge:**
Describe edge from top vertex to bottom vertex or, from left vertex to right vertex.
If both are applicable at the same time, then edge described by top vertex to bottom vertex definition dominates

**face:**
**vertex:** starting from any vertex write in clockwise direction
**edge**: follow the sequence of vertex description to write edge, i.e., in clockwise direction. 
But suppose previously, edge 7 has been defined from vertex 4 to vertex 5 using the above convention and now for maintaining the clockwise direction, edge 7 has to be described from vertex 5 to vertex 4, then write edge 7 as '-7'. 


### **One more clarification:**
Here, "[*seed
   <seed_id> <seed_x> <seed_y> <seed_z> <seed_weight>
   ...]", seed_id represents different grains; seed_x, seed_y and seed_z represents co-ordinates of center of these grains; seed_weight represents the diameter of grains to be tessellated. 
Am I correct in interpreting this?

![Neper generated image](https://github.com/user-attachments/assets/7b77aa59-8dee-4edf-8b93-7dff7b16a3b8)

**Reply by rquey** [reply]
Not exactly.  These are the coordinates and weights of the seeds from which the tessellation was generated (they do not coincide with the centroids and equivalent diameters).  However, this is of minor importance for the definition of the tess.  These data aren't used.

**Reply by Anmol-Bakshi** [reply]
Ok. Thank you.
